### PR TITLE
use go-ovn bindings for Chassis related OVN SB transactions

### PR DESCRIPTION
**Pull request by cathy-zhou**
This change replaces the ovn-sbctl calls with go-ovn bindings functions
for SB Chassis related OVN SB operations.

Signed-off-by: Yun Zhou <yunz@nvidia.com>

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/ovn/master.go [hunk 1]
```diff
@@ -880,10 +877,12 @@ func (oc *Controller) syncNodesPeriodic() {
 
 	for nodeName, chassisName := range chassisMap {
 		if chassisName != "" {
-			_, stderr, err = util.RunOVNSbctl("--if-exist", "chassis-del", chassisName)
+			cmd, err := oc.ovnSBClient.ChassisDel(chassisName)
 			if err != nil {
-				klog.Errorf("Failed to delete chassis with name %s for node %s: stderr: %s, error: %v",
-					chassisName, nodeName, stderr, err)
+				klog.Errorf("Unable to create the ChassisDel command for chassis: %s from the sbdb", chassisName)
+			} else if err = oc.ovnSBClient.Execute(cmd); err != nil {
```
Comment (abhat) [inline go-controller/pkg/ovn/master.go:883]: consider: you can also call Execute outside the for-loop. That way we only make a single transaction out to ovsdb.

File: go-controller/pkg/ovn/master.go [hunk 2]
```diff
@@ -996,63 +991,46 @@ func (oc *Controller) syncNodes(nodes []interface{}) {
 
 	for nodeName, chassisName := range chassisMap {
 		if chassisName != "" {
-			_, stderr, err = util.RunOVNSbctl("--if-exist", "chassis-del", chassisName)
+			cmd, err := oc.ovnSBClient.ChassisDel(chassisName)
 			if err != nil {
-				klog.Errorf("Failed to delete chassis with name %s for logical switch %s: stderr: %q, error: %v",
-					chassisName, nodeName, stderr, err)
+				klog.Errorf("Unable to create the ChassisDel command for chassis: %s from the sbdb", chassisName)
+			} else if err = oc.ovnSBClient.Execute(cmd); err != nil {
```
Comment (abhat) [inline go-controller/pkg/ovn/master.go:997]: same comment as in syncNodesPeriodic to batch execution of the chassisDel commands.

File: go-controller/pkg/ovn/master.go [hunk 3]
```diff
@@ -996,63 +991,46 @@ func (oc *Controller) syncNodes(nodes []interface{}) {
...
-
-func (oc *Controller) deleteNodeChassis(nodeName string) error {
-	chassisName, stderr, err := util.RunOVNSbctl("--data=bare", "--no-heading",
-		"--columns=name", "find", "Chassis",
-		"hostname="+nodeName)
-	if err != nil {
-		return fmt.Errorf("failed to get chassis name for node %s: stderr: %q, error: %v",
-			nodeName, stderr, err)
+	if len(cmds) == 0 {
+		return fmt.Errorf("failed to find chassis of node %s", nodeName)
```
Comment (abhat) [inline go-controller/pkg/ovn/master.go:1028]: nit: s/of node/for node

File: go-controller/pkg/ovn/master.go [hunk 4]
```diff
@@ -996,63 +991,46 @@ func (oc *Controller) syncNodes(nodes []interface{}) {
...
-	if chassisName == "" {
-		klog.Warningf("Chassis name is empty for node: %s", nodeName)
-	} else {
-		_, stderr, err = util.RunOVNSbctl("--if-exist", "chassis-del", chassisName)
-		if err != nil {
-			return fmt.Errorf("failed to delete chassis with name %s for node %s: stderr: %q, error: %v",
-				chassisName, nodeName, stderr, err)
-		}
+	if err = oc.ovnSBClient.Execute(cmds...); err != nil {
+		return fmt.Errorf("failed to delete chassis of name \"%s\" for node %s: error: %v",
```
Comment (abhat) [inline go-controller/pkg/ovn/master.go:1032]: nit: consider error text to be `fmt.Errorf("failed to delete chassis %s for node %s": error %v", ...)`

File: go-controller/pkg/testing/mock_chassis.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"fmt"
+
+	goovn "github.com/ebay/go-ovn"
+	"k8s.io/klog"
+)
+
+// Get logical switch port by name
+func (mock *MockOVNClient) ChassisList() ([]*goovn.Chassis, error) {
+	var ChassisCache MockObjectCacheByName
```
Comment (abhat) [inline go-controller/pkg/testing/mock_chassis.go:12]: `chassisCache` as against `ChassisCache`

File: go-controller/pkg/testing/mock_chassis.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	var ch interface{}
+	chArray := []*goovn.Chassis{}
+	for _, ch = range ChassisCache {
+		chassis, ok := ch.(*goovn.Chassis)
+		if !ok {
+			return nil, fmt.Errorf("invalid object type assertion for %s", ChassisType)
+		}
+		if chassis.Name == name {
```
Comment (abhat) [inline go-controller/pkg/testing/mock_chassis.go:60]:
```
if chassis.Name == name || chassis.Hostname == name {
        chArray = append(chArray, chassis)
}
```

File: go-controller/pkg/testing/mock_ovn.go [hunk 7]
```diff
@@ -62,12 +63,16 @@ var _ goovn.Execution = &MockExecution{}
...
 // return a new mock client to operate on db
 func NewMockOVNClient(db string) *MockOVNClient {
-	return &MockOVNClient{
+	mock := &MockOVNClient{
 		db:         db,
 		cache:      make(map[string]MockObjectCacheByName),
 		errorCache: make(map[string]error),
 		connected:  true,
 	}
+	mock.cache[LogicalSwitchPortType] = make(MockObjectCacheByName)
```
Comment (abhat) [inline go-controller/pkg/testing/mock_ovn.go:72]: why is this needed? The caller should handle `goovn.ErrorSchema`

Comment (cathy-zhou) [reply go-controller/pkg/testing/mock_ovn.go:72]: In my opinion, goovn.ErrorSchema should only be returned if the specified logical entity type is not defined in the schema. With current implementation, if no logical entity is added before ChassisList()/ChassisGet() is called, ChassisList()/Get() would return goovn.ErrorSchema, where I think the correct behavior is to return the empty list, the same as its golang binding implementation.

Comment (cathy-zhou) [reply go-controller/pkg/testing/mock_ovn.go:72]: @abhat I've addressed all your other comments and repushed the code, please take a look.  Thanks very much for all your feedback.

File: go-controller/pkg/ovn/master.go [hunk 8]
```diff
@@ -994,65 +995,54 @@ func (oc *Controller) syncNodes(nodes []interface{}) {
...
-	if chassisName == "" {
-		klog.Warningf("Chassis name is empty for node: %s", nodeName)
-	} else {
-		_, stderr, err = util.RunOVNSbctl("--if-exist", "chassis-del", chassisName)
-		if err != nil {
-			return fmt.Errorf("failed to delete chassis with name %s for node %s: stderr: %q, error: %v",
-				chassisName, nodeName, stderr, err)
-		}
+	if err = oc.ovnSBClient.Execute(cmds...); err != nil {
+		return fmt.Errorf("failed to delete chassis \"%s\" for node %s: error: %v",
```
Comment (girishmg) [inline go-controller/pkg/ovn/master.go:1044]: instead of \"%s\", you can use %q

**Comment by girishmg** [comment]
@cathy-zhou looks like syncNodes() and syncNodesPeriodic() code is very similar. Can you check if we can move them to a common function? Thanks.

File: go-controller/pkg/testing/mock_chassis.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	}, nil
+}
+
+// Get chassis by hostname or name
+func (mock *MockOVNClient) ChassisGet(name string) ([]*goovn.Chassis, error) {
+	var chassisCache MockObjectCacheByName
+	var ok bool
+	if chassisCache, ok = mock.cache[ChassisType]; !ok {
+		klog.V(5).Infof("Cache doesn't have any object of type %s", ChassisType)
+		return nil, goovn.ErrorSchema
```
Comment (abhat) [inline go-controller/pkg/testing/mock_chassis.go:50]: consider: panic here since this should never happen now. Given https://github.com/ovn-org/ovn-kubernetes/pull/1617/files#r472504580, we shouldn't return `goovn.ErrorSchema` here.

File: go-controller/pkg/testing/mock_ovn.go (hunk 7 above)
Comment (abhat) [reply go-controller/pkg/testing/mock_ovn.go:72]: Thanks. LGTM except a minor nit to panic in mock code.

File: go-controller/pkg/testing/mock_chassis.go (hunk 9 above)
Comment (cathy-zhou) [reply go-controller/pkg/testing/mock_chassis.go:50]: I am not sure. Logically it is correct if the type is not defined in the cache, it should return goovn.ErrorSchema. Also, this is a utility library function, isn't it right to return error instead of panic?

I don't have strong opinions about this though.

**Comment by girishmg** [comment]
@abhat are you fine with the latest changes?

File: go-controller/pkg/testing/mock_chassis.go (hunk 9 above)
Comment (abhat) [reply go-controller/pkg/testing/mock_chassis.go:50]: So the only reason I suggested to panic is because now when we create the mock client, we specifically populate the types in the cache. So it will be something truly amiss to not find that type in the cache. The mock_lsp code was returning `goovn.ErrorSchema` earlier because we weren't adding the types to the cache on initialization earlier. I am OK with keeping this as is. But in the past when we have had such `this should never happen` kind of errors in unit-test code, the recommendation has been to panic early.

**Comment by abhat** [comment]
There was just one comment around the object type not existing in the cache. Other than that LGTM.

File: go-controller/pkg/testing/mock_chassis.go (hunk 9 above)
Comment (cathy-zhou) [reply go-controller/pkg/testing/mock_chassis.go:50]: @abhat thanks for the explaination. I will keep it as is since you are ok with it. thanks again for all your time!

**Comment by abhat** [comment]
LGTM.
